File: LeetCode_Java/sorts/912.sort-an-array.java
package LeetCode_Java.sorts;

/*
 * @lc app=leetcode id=948 lang=java
 *
 * [912] Sort an Array
 */

// @lc code=start

import java.util.Arrays;

class Solution {
    public int[] sortArray(int[] nums) {
        quickSort(nums, 0, nums.length-1);
        return nums;
    }

    // Perform quickSort for index start to index end in nums[].
    public void quickSort(int[] nums, int start, int end) {
        if (start >= end) return;
        int pivotIdx = partition(nums, start, end);
        quickSort(nums, start, pivotIdx-1);        
        quickSort(nums, pivotIdx+1, end);
    }

    public int partition(int[] nums, int start, int end) {
        int pivotVal = nums[end];
        int pEdit = start, pSearch = start;
        while (pSearch < end) {
            if (nums[pSearch] <= pivotVal) {
                swap(nums, pEdit, pSearch);
                pEdit ++;
            }
            pSearch ++;
        }
        swap(nums, pEdit, end);
        return pEdit;
    }

    public void swap(int[] nums, int i, int j) {
        if (i == j) return;
        int temp =  nums[i];
        nums[i] = nums[j];
        nums[j] =  temp;
        return;
    }

    public static void main(String[] args) {
        int[] nums = {5,1,1,2,0,0};
        Solution sol = new Solution();
        sol.sortArray(nums);
        System.out.println(Arrays.toString(nums));
    }
}

File: LeetCode_Python/sorts/912.sort-an-array.py
#
# @lc app=leetcode id=912 lang=python3
#
# [912] Sort an Array
#

# @lc code=start
import random
class Solution:
    def swap(self, nums, i, j):
        if i == j:
            return 
        else:
            temp = nums[i]
            nums[i] = nums[j]
            nums[j] = temp
            return 
    def partition(self, nums, left, right):   # O(n)
        pivot = nums[right]
        pEdit = pSearch = left
        while (pSearch < right):
            if (nums[pSearch] <= pivot):
                self.swap(nums, pEdit, pSearch)
                pEdit+=1
            pSearch+=1
        self.swap(nums, pEdit, right)
        return pEdit


    def quickSort(self, nums, left, right):
        if left >= right: return None
        mid = self.partition(nums, left, right)
        self.quickSort(nums, left, mid-1)
        self.quickSort(nums, mid+1, right)

    def sortArray(self, nums: List[int]) -> List[int]:
        random.shuffle(nums)
        self.quickSort(nums, 0, len(nums)-1)
        return nums


        
# @lc code=end

# it is just quick sort